only attempt to restore View saved state while INITIALIZED

RBusarow opened the pull request:
fixes #570

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/workflow-kotlin?pullRequest=635) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/workflow-kotlin?pullRequest=635) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/workflow-kotlin?pullRequest=635) it.</sub>

steve-the-edwards commented on workflow-ui/container-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt:
```diff
@@ -50,9 +51,14 @@ internal constructor(
+        // We're only allowed to restore from an INITIALIZED state, but this callback can also be
+        // invoked while the owner is already CREATED.
+        // https://github.com/square/workflow-kotlin/issues/570
+        ?.takeIf { it.lifecycle.currentState == Lifecycle.State.INITIALIZED }
```
I think @rjrjr originally tried this solution but decided against it? I can't remember now.

RBusarow replied:
I believe it was just a matter of being able to reproduce the bug first.

If this line is removed, everything still passes except for the new test, even if you call `scenario.recreate()`, or change orientation or something.

rjrjr replied:
Yup, I feel fine about the fix, just didn't want to do it w/o proof.

steve-the-edwards reviewed (approved):
Is this just pulling this fix into main from the ui-update branch?

rjrjr commented on workflow-ui/container-android/src/androidTest/java/com/squareup/workflow1/ui/backstack/test/BackstackContainerTest.kt:
```diff
@@ -60,6 +62,59 @@ internal class BackstackContainerTest {
+      it.nestedTestView().viewState = "some state"
+    }
+
+    // Recreating the activity sends one ONE_CREATE event,
```
🤩

rjrjr commented:
@RBusarow what changed in those last couple of force pushes?

RBusarow commented:
> @RBusarow what changed in those last couple of force pushes?

I fixed this typo: 
![image](https://user-images.githubusercontent.com/9890900/150232396-b33f5e8a-18ad-4a75-85c1-4a1329d6ab58.png)

And I added that same `.api` file change.

rjrjr commented:
@RBusarow When I revert everything but the test from this, the test still passes. Do you remember if there was any particular AVD or API level you were using to reproduce the failure?

RBusarow commented:
@rjrjr I just tested against a 31 "Pixel 5" AVD and with just the `INITIALIZED` check commented out it failed.  I then ran it against API 29 and it passed.

rjrjr commented:
Perfect, thanks.

I don't know that these two issues are actually related, still debugging.

rjrjr commented:
Seems unrelated, just a missing call to `restoreRegistryControllerIfReady` in `ModalContainer`. Phew!
